crates/matrix-sdk-indexeddb/src/state_store/migrations.rs

```diff
@@ -111,65 +110,72 @@ pub async fn upgrade_inner_db(
-        let created = Arc::new(AtomicBool::new(false));
-        let created_inner = created.clone();
-
-        db_req.set_on_upgrade_needed(Some(
-            move |evt: &IdbVersionChangeEvent| -> Result<(), JsValue> {
-                // in case this is a fresh db, we dont't want to trigger
-                // further migrations other than just creating the full
-                // schema.
-                if evt.old_version() < 1.0 {
-                    create_stores(evt.db())?;
-                    created_inner.store(true, Ordering::Relaxed);
-                }
-                Ok(())
-            },
-        ));
-
-        let pre_db = db_req.into_future().await?;
-        let old_version = pre_db.version();
+        let pre_db = IdbDatabase::open(name)?.into_future().await?;
+
+        // Even if the web-sys bindings expose the version as a f64, the IndexedDB API
+        // works with an unsigned integer.
+        // See <https://github.com/rustwasm/wasm-bindgen/issues/1149>
+        let mut old_version = pre_db.version() as u32;
+
+        // Inside the `onupgradeneeded` callback we would know whether it's a new DB
+        // because the old version would be set to 0, here it is already set to 1 so we
+        // check if the stores exist.
+        if old_version == 1 && pre_db.object_store_names().next().is_none() {
+            old_version = 0;
+        }
 
-        if created.load(Ordering::Relaxed) {
-            // this is a fresh DB, nothing to do
-        } else if old_version == 1.0 && has_store_cipher {
+        if old_version == 0 {
+            should_create_stores = true;
+            // The stores are brand new, other migrations are unnecessary.
+        } else if old_version < 2 && has_store_cipher {
             match migration_strategy {
                 MigrationConflictStrategy::BackupAndDrop => {
-                    backup(&pre_db, meta_db).await?;
-                    recreate_stores = true;
+                    backup_v1(&pre_db, meta_db).await?;
+                    should_drop_stores = true;
+                    should_create_stores = true;
                 }
                 MigrationConflictStrategy::Drop => {
-                    recreate_stores = true;
+                    should_drop_stores = true;
+                    should_create_stores = true;
                 }
                 MigrationConflictStrategy::Raise => {
                     return Err(IndexeddbStateStoreError::MigrationConflict {
                         name: name.to_owned(),
                         old_version,
                         new_version: CURRENT_DB_VERSION,
-                    })
+                    });
                 }
             }
-        } else if old_version < 1.2 {
-            migrate_to_v1_2(&pre_db, store_cipher).await?;
-        } else {
-            // Nothing to be done
+            // The stores are brand new, other migrations are unnecessary.
```

Comment:
I'm failing to understand what this comment tries to say here. Is it about the `else if` below not being a regular `if`?

Reply:
Yes, both comments are about that. I should probably put them on the line of the `else if`?

Reply:
I'm not sure. I would probably put a comment about the `else if` in the same place. Maybe remove the duplicate comments (that also look a bit like a copy-paste mistake) and instead add one comment on top of the whole `if` / `else if` chain that says sth. like "Using exclusive branches because in the case of version 0 or 1, …".